fmc/src/flow/pcr.rs

```diff
@@ -56,16 +77,62 @@ pub fn extend_journey_pcr(env: &mut FmcEnv, hand_off: &HandOff) -> CaliptraResul
 /// * `env` - FMC Environment
 /// * `pcr_id` - PCR slot to extend the data into
 fn extend_pcr_common(env: &mut FmcEnv, hand_off: &HandOff, pcr_id: PcrId) -> CaliptraResult<()> {
+    let manifest_digest = Tci::image_manifest_digest(env, hand_off);
+    let manifest_digest: [u8; 48] = okref(&manifest_digest)?.into();
+
     // Extend RT TCI (Hash over runtime code)
     let rt_tci = Tci::rt_tci(env, hand_off);
-    let rt_tci: [u8; 48] = okref(&rt_tci)?.into();
-    env.pcr_bank.extend_pcr(pcr_id, &mut env.sha384, &rt_tci)?;
+    let rt_tci_bytes: [u8; 48] = okref(&rt_tci)?.into();
+
+    let mut extender = PcrExtender {
+        pcr_bank: &mut env.pcr_bank,
+        sha384: &mut env.sha384,
+    };
+
+    extender.extend(pcr_id, &rt_tci_bytes, PcrLogEntryId::RtTci)?;
 
     // Extend FW Image Manifest
-    let manifest_digest = Tci::image_manifest_digest(env, hand_off);
-    let manifest_digest: [u8; 48] = okref(&manifest_digest)?.into();
-    env.pcr_bank
-        .extend_pcr(pcr_id, &mut env.sha384, &manifest_digest)?;
+    extender.extend(pcr_id, &manifest_digest, PcrLogEntryId::ManifestDigest)?;
+
+    Ok(())
+}
+
+/// Log PCR data
+///
+/// # Arguments
+/// * `pcr_entry_id` - PCR log entry ID
+/// * `pcr_id` - PCR ID
+/// * `data` - PCR data
+///
+/// # Return Value
+/// * `Ok(())` - Success
+/// * `Err(GlobalErr::PcrLogInvalidEntryId)` - Invalid PCR log entry ID
+/// * `Err(GlobalErr::PcrLogUpsupportedDataLength)` - Unsupported data length
+///
+pub fn log_pcr(pcr_entry_id: PcrLogEntryId, pcr_id: PcrId, data: &[u8]) -> CaliptraResult<()> {
+    if pcr_entry_id == PcrLogEntryId::Invalid {
+        return Err(CaliptraError::FMC_PCR_LOG_INVALID_ENTRY_ID);
+    }
+
+    if data.len() > 48 {
+        return Err(CaliptraError::FMC_PCR_LOG_UNSUPPORTED_DATA_LENGTH);
+    }
+
+    // Create a PCR log entry
+    let mut pcr_log_entry = PcrLogEntry {
+        id: pcr_entry_id as u16,
+        pcr_id: pcr_id as u16,
+        ..Default::default()
+    };
+    pcr_log_entry.pcr_data.as_bytes_mut()[..data.len()].copy_from_slice(data);
+
+    let dst: &mut [PcrLogEntry] = unsafe {
+        let ptr = PCR_LOG_ORG as *mut PcrLogEntry;
+        core::slice::from_raw_parts_mut(ptr, PCR_LOG_SIZE / core::mem::size_of::<PcrLogEntry>())
+    };
+
+    // Store the log entry.
+    dst[pcr_entry_id as usize - 1] = pcr_log_entry;
```

Comment:
This current logic will leave holes in the log buffer if ROM and FMC do not agree on how many extensions should have occurred. https://github.com/chipsalliance/caliptra-sw/pull/482 brings in a log entry counter, which ROM initializes to zero, and which FMC can initialize by walking the log space until it finds an invalid entry. Or, could have ROM put the entry count in FHT.

Reply:
I think it would make the most sense for ROM to fill in this count, and for FMC to update the count